test/unit/puppet_fact_parser_test.rb

```diff
@@ -233,8 +234,35 @@ def setup
     assert_equal '172.30.43.0', parser.interfaces['local_area_connection_2'][:network]
   end
 
+  test "should create connection between new os and new arch on facts parsing" do
+    assert os_to_arch_connectivity_check
+  end
+
+  test "should create connection between existing os and new arch on facts parsing" do
+    FactoryGirl.create(:operatingsystem,:name => 'AnyOS', :major => '6')
+    assert os_to_arch_connectivity_check
```

Comment:
@domcleal IMHO its better to keep it in the fact parser, as its the fact parser responsibility is to create objects prior to the host create/update.
i would prefer to reduce the responsibility of the host class, also just in case it might fail from what ever reason, the relationship between os/arch can still be established.